File: Algorithm/2023.11/PakaOxO/1,2,3떨어트리기.js
/**
 * Programmers_1,2,3 떨어트리기
 *  - 문제 분류: 트리
 */
const solution = (edges, target) => {
  /* 변수 관리 */
  const N = edges.length;
  const count = Array.from({ length: N + 1 }, () => 0);
  const children = Array.from({ length: N + 1 }, () => []);
  let visit = [];
  let answer = [];

  /* 메인 로직 */
  for (let i=0; i<edges.length; i++) {
    const [parent, child] = edges[i];
    // 노드에 자식노드 정보 추가
    children[parent - 1].push(child - 1);
  }
  for (let i=0; i<=N; i++) children[i].sort((a, b) => a - b);

  while (true) {
    const flag = check(); // 리프노드 방문개수 유효성 체크
    if (flag < 0) { // target이 0인데 방문 되었을 경우
      visit = null;
      break;
    }
    if (flag > 0) break; // 조건에 맞는 최소 방문 횟수를 찾음

    let leaf = 0; // 방문 리프노드를 찾아보자
    while (children[leaf].length > 0) {
      const len = children[leaf].length;
      leaf = children[leaf][count[leaf]++ % len];
    }
    count[leaf]++;
    visit.push(leaf);
  }

  let pointer = 0; // 리프노드에 삽입할 숫자를 정해보자
  while (visit !== null && pointer < visit.length) {
    const leaf = visit[pointer++];
    for (let i=1; i<=3; i++) {
      if ((target[leaf] - i) > (count[leaf] - 1) * 3) continue; // 이 숫자를 넣으면 target 못 채움
      answer.push(i);
      target[leaf] -= i;
      count[leaf]--;
      break;
    }
  }

  if (answer.length === 0) answer.push(-1);
  /* 정답 반환 */
  return answer;

  // check
  function check() {
    for (let i=0; i<=N; i++) {
      if (children[i].length > 0) continue;
      if (target[i] === 0 && count[i] > 0) return -1;
      if (target[i] < count[i]) return -1;
      if (!(target[i] >= count[i] && target[i] <= count[i] * 3)) return 0;
    }
    return 1;
  }
}

console.log(solution([
  [2, 4], [1, 2], [6, 8], [1, 3], [5, 7], [2, 5], [3, 6], [6, 10], [6, 9]
],	[0, 0, 0, 3, 0, 0, 5, 1, 2, 3]));
console.log(solution([[1, 2], [1, 3]],	[0, 7, 3]));
console.log(solution([[1, 3], [1, 2]],	[0, 7, 1]));

File: Algorithm/2023.11/BinaryHyeok/1,2,3 떨어트리기.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges, int[] target) { 
        // tree 초기화
        int n = edges.length + 1;
        List<Integer>[] tree = new ArrayList[n];
        for(int i = 0; i < n; i++){
            tree[i] = new ArrayList<>();
        }
        
        // tree 값 입력
        for(int i = 0; i < edges.length; i++){
            int p = edges[i][0];
            int c = edges[i][1];
            // 넣어줄 때 배열 인덱스에 맞춰서 저장
            tree[p - 1].add(c - 1);
        }
        
        int TC = 0;
        // tree 오름차순 정렬
        for(int i = 0; i < n; i++){
            Collections.sort(tree[i]);
            // 해당 노드가 리프노드이면서 타겟이 있다면 숫자를 떨어뜨려야 하는 노드
            if(tree[i].isEmpty() && target[i] > 0){
                TC++;
            }
        }
        
        // 현재 노드가 몇 번 지나갔는지 체크
        int[] pass = new int[n];
        // 현재 노드가 보유하고 있는 숫자 개수
        int[] count = new int[n];
        // 방문 노드인지 저장
        boolean[] visited = new boolean[n];
        // 리프 노드 저장
        List<Integer> leafNode = new ArrayList<>();

        while(TC > 0){
            // 루트노드
            int node = 0;
            
            // 숫자떨어뜨리면서 리프노드까지 내려가기
            while(tree[node].size() > 0){
                /*
                    노드의 순서를 정하는 방법
                    노드의 작은 순서대로 이동 -> 정렬해놨으니 0번부터 시작하면된다.
                    방문횟수 % 자식노드의 수 -> 0 ~ 자식노드의 횟수 - 1 까지 차례대로 나옴
                */
                node = tree[node].get(pass[node]++ % tree[node].size());
            }
            
            // 리프 노드에 떨어진 숫자 증가
            count[node]++;
            leafNode.add(node);
            
            // 리프 노드의 숫자가 모두 1이라고 할 때, target보다 크다면 조건 만족 X
            if(count[node] > target[node]){
                return new int[]{-1};
            }
            
            // 리프노드를 방문하지 않았고, 노드의 총 합이 target보다 크다면 해당 노드는 만족 O
            // 개수 * 3이 타겟보다 크다면 해당 리프노드는 target 값을 만족 O
            if(!visited[node] && count[node] * 3 >= target[node]){
                visited[node] = true;
                TC--;
            }
        }
        
        // 정답을 넣을 배열
        List<Integer> result = new ArrayList<>();
        
        // 리프노드를 하나씩 꺼내서 숫자 대입
        for(int i = 0; i < leafNode.size(); i++){
            int num = leafNode.get(i);
            count[num]--;
            
            for(int j = 1; j <= 3; j++){
                // 개수 * 1이 target - j 보다 작거나 같고, 개수 * 3이 target - j보다 클 경우 target값 만족 가능
                if(count[num] <= target[num] - j && count[num] * 3 >= target[num] - j){
                    target[num] -= j;
                    result.add(j);
                    break;
                }
            }
        }
        
        // List -> Array
        int[] answer = new int[result.size()];
        for(int i = 0; i < result.size(); i++){
            answer[i] = result.get(i);
        }
        
        return answer;
    }
}